### DISCUSSION TITLE
How to update user account??

### DISCUSSION THREAD

**Discussion by anishmishra0** [GodotNuts/GodotFirebase#423] [Q&A] [answered]
I want to update Display name and photo of the user. I saw a func in auth.gd but this only changes Display name, photo is not changing

```
# User Profile handlers
func update_account(idToken : String, displayName : String, photoUrl : String, deleteAttribute : PackedStringArray, returnSecureToken : bool) -> void:
	if _is_ready():
		is_busy = true
		_update_profile_body.idToken = idToken
		_update_profile_body.displayName = displayName
		_update_profile_body.photoUrl = photoUrl
		_update_profile_body.deleteAttribute = deleteAttribute
		_update_profile_body.returnSecureToken = returnSecureToken
		var err = request(_base_url + _update_account_request_url, _headers, HTTPClient.METHOD_POST, JSON.stringify(_update_profile_body))
		if err != OK:
			is_busy = false
			Firebase._printerr("Error updating account: %s" % err)
		else:
			print("done")
```

**Comment by WolfgangSenff** [comment]
Hm. I think I see what's happening. When the user's account info is updated, we don't automatically get the updated values - it's not sent back to the client like that, I'm guessing. You can try this as a way to fix it for now (I'll consider adding something to update the local values) - when you call that function, do it this way (formatting here sucks, sorry):

```
Firebase.Auth.update_account(variables_you_use)
yield(get_tree().create_timer(0.5), "timeout")
Firebase.Auth.get_user_data()
var user = yield(Firebase.Auth, "userdata_received") # Yield to which I'm referring below
print(user)
```

In theory, the above should update the user and return it. If you don't want to use that yield, you can just connect to `userdata_received` with an actual function that handles it correctly and that'll also work.

**Reply by anishmishra0** [reply]
I am exactly doing that but after getting userdata_received photourl is not returned.
This is my function
```
func update_user_account():
	Firebase.Auth.update_account(Firebase.Auth.auth.idtoken,"Display_Name","https://github.com/anishmishra0/Warrior-Game-developement-in-python/blob/master/part2/img/player.png?raw=true",PackedStringArray(),true)
	await get_tree().create_timer(3).timeout
	Firebase.Auth.get_user_data()
	var data = await Firebase.Auth.userdata_received
	print(data)
```
Output:
![Screenshot 2024-07-31 170709](https://github.com/user-attachments/assets/879a0340-a875-4297-82b1-894bc8511fb0)

**Comment by WolfgangSenff** [comment, marked as answer]
Ah, dangit, I see why. I can tell you how to fix it locally and I can try to fix it eventually, but I'm fairly busy right now so can't implement what I know to be a fix for a little while. Here's what I'd do:

1. Go into this file in your local version of the addon: https://github.com/GodotNuts/GodotFirebase/blob/main/addons/godot-firebase/auth/user_data.gd
2. Go to line 32 (`photo_url = provider_user_info[0].get("photoUrl", "")`) and copy it, then paste it above the if statement above it and unindent it one tab
3. On the new line you just added, change `provider_user_info[0]` to just be `p_userdata`

In theory, that should work. The other issue you'll see though is if you look in the _to_string below it, it doesn't include the photo_url, so you wouldn't physically see it when printing it out. You can add it easily though, and I will do so when I have time to implement the above and make sure it works. Let me know if it works for you, and that would make it get implemented by me faster. :P

**Reply by anishmishra0** [reply]
Yes the problem is solved now

> The other issue you'll see though is if you look in the _to_string below it, it doesn't include the photo_url, so you wouldn't physically see it when printing it out.

I think the problem was this all along. I just added `txt += "photo url : %s\n" % photo_url` in _to_string() func and url is now printed

**Reply by WolfgangSenff** [reply]
Oh really? That's odd. It's definitely not being set where I would expect it to, but if the change to the printing works alone, then that's a better and easier fix!

**Reply by anishmishra0** [reply]
I need another advice
Should I upload the photo of the user to the storage and then get back the link and save it to photo_url. Or i can do something better??

**Reply by WolfgangSenff** [reply]
You could certainly do that, yeah. The reason the photoUrl is a bit odd is because it normally comes when you have a provider - like Facebook - supply the data. The Google API will automatically add the photoUrl if they login with Facebook, for instance, to point to their profile pic, I believe. I'd probably do what you're suggesting, though, if they're allowed to upload a photo from your app. I don't really recommend doing that though unless you're very proficient in JavaScript, for the reason that I don't like users being able to upload dirty pictures as their profile pics. I have a way to work around it, but it costs money (using Google's own infrastructure to scan the image to ensure it's safe to use in any general application, even if kid's use it).

**Reply by anishmishra0** [reply]
It's a personal budgeting application and user will only able to upload one profile picture.
BTW users cannot even interect or see profile pic of other user.  So I think i will go with this method

**Reply by WolfgangSenff** [reply]
Oh okay, well that should work just fine then!
